core/include/traccc/fitting/kalman_filter/gain_matrix_smoother.hpp

```diff
@@ -44,67 +44,117 @@ struct gain_matrix_smoother {
-        const matrix_type<6, 6>& cur_filtered_cov = cur_filtered.covariance();
-
-        // Regularization matrix for numerical stability
-        static constexpr scalar_type epsilon = 1e-13f;
-        const matrix_type<6, 6> regularization =
-            matrix_operator().template identity<e_bound_size, e_bound_size>() *
-            epsilon;
-        const matrix_type<6, 6> regularized_predicted_cov =
-            next_predicted_cov + regularization;
-
-        // Calculate smoothed parameter for current state
-        const matrix_type<6, 6> A =
-            cur_filtered_cov * matrix_operator().transpose(next_jacobian) *
-            matrix_operator().inverse(regularized_predicted_cov);
-
-        const matrix_type<6, 1> smt_vec =
-            cur_filtered_vec + A * (next_smoothed_vec - next_predicted_vec);
-        const matrix_type<6, 6> smt_cov =
-            cur_filtered_cov + A * (next_smoothed_cov - next_predicted_cov) *
-                                   matrix_operator().transpose(A);
-
-        cur_state.smoothed().set_vector(smt_vec);
-        cur_state.smoothed().set_covariance(smt_cov);
-
-        // projection matrix
-        // @TODO: Need to check if it is right to put the filtered state here
-        const typename mask_group_t::value_type::projection_matrix_type H =
-            mask_group[index].projection_matrix(cur_filtered);
-
-        constexpr const unsigned int D =
-            mask_group_t::value_type::shape::meas_dim;
-
-        // Calculate smoothed chi square
-        const matrix_type<D, 1>& meas_local = cur_state.measurement_local();
-        const matrix_type<D, D>& V = cur_state.measurement_covariance();
-        const matrix_type<D, 1> residual = meas_local - H * smt_vec;
-        const matrix_type<D, D> R =
-            V - H * smt_cov * matrix_operator().transpose(H);
-        const matrix_type<1, 1> chi2 = matrix_operator().transpose(residual) *
-                                       matrix_operator().inverse(R) * residual;
-
-        cur_state.smoothed_chi2() = matrix_operator().element(chi2, 0, 0);
+        this->operator()(cur_state, next_state);
     }
+
+    template <size_type D = 2u>
+    TRACCC_HOST_DEVICE inline void operator()(
+        track_state<algebra_t>& cur_state,
+        const track_state<algebra_t>& next_state) const {
+        static_assert(D <= 2u,
+                      "The measurement dimension should be less or equal to 2");
+
+        gain_matrix_smoother_helper<D>()(cur_state, next_state);
+    }
+
+    template <size_type D, typename Enable = void>
+    struct gain_matrix_smoother_helper;
+
+    template <size_type D>
+    struct gain_matrix_smoother_helper<D, typename std::enable_if_t<D == 0u>> {
+
+        TRACCC_HOST_DEVICE inline void operator()(
+            track_state<algebra_t>& /*cur_state*/,
+            const track_state<algebra_t>& /*next_state*/) const {
+            return;
+        }
+    };
```

Comment:
The suggested change doesn't seem to work with the specilation of `0u`:

```
In file included from /home/beomki/projects/traccc/traccc/core/include/traccc/fitting/kalman_filter/kalman_actor.hpp:13,
                 from /home/beomki/projects/traccc/traccc/core/include/traccc/fitting/kalman_filter/kalman_fitter.hpp:17,
                 from /home/beomki/projects/traccc/traccc/tests/common/tests/kalman_fitting_test.hpp:12,
                 from /home/beomki/projects/traccc/traccc/tests/common/tests/kalman_fitting_test.cpp:16:
/home/beomki/projects/traccc/traccc/core/include/traccc/fitting/kalman_filter/gain_matrix_updater.hpp:142:15: error: explicit specialization in non-namespace scope ‘struct traccc::gain_matrix_updater<algebra_t>’
  142 |     template <>
      |               ^
/home/beomki/projects/traccc/traccc/core/include/traccc/fitting/kalman_filter/gain_matrix_updater.hpp:143:12: error: template parameters not deducible in partial specialization:
  143 |     struct gain_matrix_updater_helper<0u> {
```

[There is a way to bypass the error](https://stackoverflow.com/questions/3052579/explicit-specialization-in-non-namespace-scope) but it already makes things too complicated.. :p

Reply:
I think the specialization for ```D==0``` is needed because ```gain_matrix_smoother_helper``` is recursive. But would a simple switch-case not have done the trick?
```
switch(meas.meas_dim) {
case 1: gain_matrix_smoother_helper<1>()(cur_state, next_state); break;
case 2: gain_matrix_smoother_helper<2>()(cur_state, next_state); break;
case 0: ERROR
};
```
Then ```gain_matrix_smoother_helper``` can even be a simple function